docs/multiplexing: overhaul plugin documentation (#14509)
hashicorp/vault: website/content/docs/internals/plugins/plugin-development.mdx

---
layout: docs
page_title: Plugin Development
description: Learn about Vault plugin development.
---

# Plugin Development

~> Advanced topic! Plugin development is a highly advanced topic in Vault, and
is not required knowledge for day-to-day usage. If you don't plan on writing any

> plugins, we recommend not reading this section of the documentation.

Reviewer: ```suggestion
plugins, feel free to skip this section of the documentation.
```

Because Vault communicates to plugins over a RPC interface, you can build and
distribute a plugin for Vault without having to rebuild Vault itself. This makes
it easy for you to build a Vault plugin for your organization's internal use,
for a proprietary API that you don't want to open source, or to prototype
something before contributing it back to the main project.

In theory, because the plugin interface is HTTP, you could even develop a plugin

> using a completely different programming language! (Disclaimer, you would also

Reviewer: ```suggestion
using a completely different programming language. Please note that with plugin development, you must re-implement the plugin API. This effort may be extremely-time consuming and requires a lot of work. 
```

> have to re-implement the plugin API which is not a trivial amount of work.)

Reviewer: ```suggestion
)
```
